### PR TITLE
Distribute the broken chunks

### PR DESCRIPTION
**Pull request by AymericDu**
##### SUMMARY

- Improve log messages of all rebuilders to better track progress
- `oio-blob-rebuilder`: Distribute broken chunks to several beanstalk

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

- `oio-meta1-rebuilder`
- `oio-meta2-rebuilder`
- `oio-blob-rebuilder`

##### SDS VERSION

```
openio 4.2.2.dev145
```

### CODE REVIEW

File: oio/rebuilder/blob_rebuilder.py [hunk 1]
```diff
@@ -33,211 +36,293 @@
...
+    def _fetch_chunks(self, **kwargs):
+        if self.input_file:
+            return self._fetch_chunks_from_file(**kwargs)
+        if self.beanstalkd_addr and not self.distributed_addr:
+            return self._fetch_events_from_beanstalk(**kwargs)
+        if self.volume:
+            return self.rdir_client.chunk_fetch(
+                self.volume, limit=self.rdir_fetch_limit, rebuild=True,
+                **kwargs)
+        raise ConfigurationException('No source to fetch chunks')
```
Comment (fvennetier) [inline oio/rebuilder/blob_rebuilder.py]: fetch chunks *from*

File: oio/rebuilder/blob_rebuilder.py [hunk 2]
```diff
@@ -33,211 +36,293 @@
...
+            else:
+                success = super(BlobRebuilder, self).rebuilder_pass(**kwargs)
+        finally:
+            if self.volume:
+                self.rdir_client.admin_unlock(self.volume)
+        return success
+
+    def distributed_rebuilder_pass(self, **kwargs):
+        if not self.beanstalkd_addr:
+            raise ConfigurationException("No beanstalkd to fetch responses")
```
Comment (fvennetier) [inline oio/rebuilder/blob_rebuilder.py]: fetch responses *from*

File: oio/rebuilder/blob_rebuilder.py [hunk 3]
```diff
@@ -271,15 +356,122 @@ def chunk_rebuild(self, container_id, content_id, chunk_id_or_pos):
...
+                                  self.addr, self.tube)
+                self._connect(**kwargs)
+
+            job_id, data = self.beanstalkd.reserve()
+            for event_info in on_event(job_id, data, **kwargs):
+                yield event_info
+            self.beanstalkd.delete(job_id)
+        except ConnectionError as exc:
+            self.connected = False
+            self.logger.debug('Disconnected to %s using tube %s (job=%s): %s',
```
Comment (fvennetier) [inline oio/rebuilder/blob_rebuilder.py]: Disconnected *from*

File: oio/rebuilder/blob_rebuilder.py [hunk 4]
```diff
@@ -271,15 +356,122 @@ def chunk_rebuild(self, container_id, content_id, chunk_id_or_pos):
...
+
+            with self.lock_nb_events:
+                job_id = self.beanstalkd.put(event)
+                self.nb_events += 1
+                if self.nb_events == self.limit:
+                    self.fill = False
+            return True
+        except ConnectionError as exc:
+            self.connected = False
+            self.logger.debug('Disconnected to %s using tube %s (job=%s): %s',
```
Comment (fvennetier) [inline oio/rebuilder/blob_rebuilder.py]: Disconnected *from*

File: oio/rebuilder/blob_rebuilder.py [hunk 5]
```diff
@@ -33,211 +36,293 @@
...
+            event = self._event_from_broken_chunk(
+                broken_chunk, callback, **kwargs)
+            while True:
+                for _ in range(n):
+                    success = senders[index].send_event(
+                        event, **kwargs)
+                    index = (index + 1) % n
+                    if success:
+                        return index
+                time.sleep(5)
```
[the remark this answers is not in the document]
Comment (AymericDu) [reply oio/rebuilder/blob_rebuilder.py]: Each sender has an event limit. If the sender is full, `send_event` returns `False` (if there is an error, write a log message and return `False` too). So, if all senders are full, we wait 5 seconds to not use the CPU unnecessarily. As soon as a sender is no longer full, `send_event` returns `True` and we exit this loop.

File: oio/rebuilder/blob_rebuilder.py [hunk 6]
```diff
@@ -34,210 +37,302 @@ class BlobRebuilder(Rebuilder):
...
+                                         DEFAULT_REBUILDER_TUBE)
+        self.sending = False
+
+    def _rebuilder_pass(self, **kwargs):
+        if not self.beanstalkd_addr:
+            raise ConfigurationException(
+                "No beanstalkd to fetch responses from")
+        if self.beanstalkd_tube == self.distributed_tube:
+            raise ConfigurationException(
+                "The beanstalkd tubes must be different")
```
Comment (fvennetier) [inline oio/rebuilder/blob_rebuilder.py]: These checks should be done in class' constructor.

**Pull request merged**